src/gpu/lib.ts

```diff
@@ -0,0 +1,190 @@
+import { GPU } from 'gpu.js'
+import { TypeError } from '../utils/rttc'
+import { isArray } from 'util'
+
+// helper function to build 2D array output
+function buildArray(arr: Float32Array[][], end: any, res: any) {
+  for (let i = 0; i < end[0]; i++) {
+    res[i] = prettyOutput(arr[i])
+  }
+}
+
+function build2DArray(arr: Float32Array[][], end: any, res: any) {
+  for (let i = 0; i < end[0]; i++) {
+    for (let j = 0; j < end[1]; j++) {
+      res[i][j] = prettyOutput(arr[i][j])
+    }
+  }
+}
+
+// helper function to build 3D array output
+function build3DArray(arr: Float32Array[][][], end: any, res: any) {
+  for (let i = 0; i < end[0]; i++) {
+    for (let j = 0; j < end[1]; j++) {
+      for (let k = 0; k < end[2]; k++) {
+        res[i][j][k] = prettyOutput(arr[i][j][k])
+      }
+    }
+  }
+}
+
+function prettyOutput(arr: any): any {
+  if (!(arr instanceof Float32Array)) {
+    return arr
+  }
+
+  const res = arr.map(x => prettyOutput(x))
+  return Array.from(res)
+}
+
+// helper function to check array is initialized
+function checkArray(arr: any): boolean {
+  return Array.isArray(arr)
+}
+
+// helper function to check 2D array is initialized
+function checkArray2D(arr: any, end: any): boolean {
+  for (let i = 0; i < end[0]; i = i + 1) {
+    if (!Array.isArray(arr[i])) return false
+  }
+  return true
+}
+
+// helper function to check 3D array is initialized
+function checkArray3D(arr: any, end: any): boolean {
+  for (let i = 0; i < end[0]; i = i + 1) {
+    if (!Array.isArray(arr[i])) return false
+    for (let j = 0; j < end[1]; j = j + 1) {
+      if (!Array.isArray(arr[i][j])) return false
+    }
+  }
+  return true
+}
+
+function checkForNumber(arr: any): boolean {
+  if (isArray(arr)) {
+    return arr.length > 0 && arr.filter(x => !checkForNumber(x)).length === 0
+  }
+
+  return typeof arr === 'number'
+}
+
+/*
+ * we only use the gpu if:
+ * 1. we are working with numbers
+ * 2. we have a large array (> 100 elements)
+ */
+function checkValidGPU(f: any, end: any): boolean {
+  let res: any
+  if (end.length === 1) res = f(0)
+  if (end.length === 2) res = f(0, 0)
+  if (end.length === 3) res = f(0, 0, 0)
+
+  if (typeof res !== 'number') {
+    if (!Array.isArray(res)) {
+      return false
+    }
+
+    if (!checkForNumber(res)) {
+      return false
+    }
+  }
+
+  let cnt = 1
+  for (const i of end) {
+    cnt = cnt * i
+  }
+
+  return cnt > 100
```

Comment:
Shall we make the threshold a constant in some file?

Reply:
yup that's a good idea! We will put it on top of this file for now and update it again later